crates/aptos-speculative/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+// Suppress incorrect warning, as without mut the take_counts method stops compiling.
+#![allow(unused_mut)]
+
+use aptos_infallible::Mutex;
+use crossbeam::utils::CachePadded;
+use once_cell::sync::OnceCell;
+use rayon::prelude::*;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+#[cfg(test)]
+mod tests;
+
+/// This crate is designed to help with tracking events and state that might be speculative in
+/// nature due to speculative execution of transactions (e.g. by BlockSTM parallel executor, but
+/// also in any other context). The idea is that a transaction may be speculative executed, but
+/// then the this speculative execution might be invalidated and discarded, possibly triggering
+/// another re-execution. In this case, it is convenient to buffer certain state and events and
+/// also discard them, while having a way to flush in case the execution is actually finalized.
+/// All components here can (and are intended to) be used in a concurrent fashion.
+///
+/// An important feature is that flush for events happens asynchronously (and in parallel via
+/// rayon global pool), and a number of optimizations ensures that the operations on the critical
+/// path do not introduce too much overhead (e.g. clearing, initialization).
+///
+/// As far as the state is concerned, the crate currently just implements a SpeculativeCounter.
+/// In the future, we could easily implement a structure similar to SpeculativeEvents to
+/// keep track of any speculative state, where a trait could allow updating it (running state).
+
+// A struct that allows storing elements of type T as in a Vec and at any moment clearing
+// the stored elements but only in a lazy / logical manner - i.e. the elements are still
+// stored, but ignored. Useful to defer costs off the critical path.
+struct LazyClearVec<T> {
+    // Stores the elements, including some that might be logically cleared.
+    elements: Vec<T>,
+    // Logically recording the size of the logically cleared prefix.
+    skip_cleared: usize,
```

Comment:
How do we prevent this from causing memory leaks? Also, is it really that slow to clear items? i.e., why doesn't `clear()` just call `self.elements = vec![]`?

Reply:
thanks, this was super dumb. Just doing clear is fine, and we can optimize the vector to have the correct allocation behavior when we start using the crate (depending on what's the usual size of the buffer).